lib/android/app/src/main/java/com/reactnativenavigation/views/stack/topbar/titlebar/TitleAndButtonsContainer.kt

```diff
@@ -0,0 +1,181 @@
+        this.addView(titleSubTitleBar, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.MATCH_PARENT))
+        this.addView(rightButtonBar, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.MATCH_PARENT))
+    }
+
+    fun setComponent(component: View, alignment: Alignment = Alignment.Default) {
+        if (this.component == component) return
+        clearCurrentTitle()
+        this.component = component
+        this.addView(this.component, LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT))
+        titleComponentAlignment = alignment
+    }
+
+    fun setTitle(title: CharSequence?) {
+        clearComponent()
+        this.titleSubTitleBar.visibility = View.VISIBLE
+        this.titleSubTitleBar.setTitle(title)
+    }
+
+    fun setSubtitle(title: CharSequence?) {
+        clearComponent()
+        this.titleSubTitleBar.visibility = View.VISIBLE
+        this.titleSubTitleBar.setSubtitle(title)
+    }
+
+    fun setTitleBarAlignment(alignment: Alignment) {
+        titleComponentAlignment = alignment
+    }
+
+    override fun setLayoutDirection(layoutDirection: Int) {
+        super.setLayoutDirection(layoutDirection)
+        this.component?.layoutDirection = layoutDirection
+        this.titleSubTitleBar.layoutDirection = layoutDirection
+        this.rightButtonBar.layoutDirection = layoutDirection
+        this.leftButtonBar.layoutDirection = layoutDirection
+    }
+
+    fun setSubTitleTextAlignment(alignment: Alignment) = this.titleSubTitleBar.setSubTitleAlignment(alignment)
+
+    fun setTitleTextAlignment(alignment: Alignment) = this.titleSubTitleBar.setTitleAlignment(alignment)
+
+    fun setBackgroundColor(color: Colour) = if (color.hasValue()) setBackgroundColor(color.get()) else Unit
+
+    fun setTitleFontSize(size: Float) = this.titleSubTitleBar.setTitleFontSize(size)
+
+    fun setTitleTypeface(typefaceLoader: TypefaceLoader, font: FontOptions) = this.titleSubTitleBar.setTitleTypeface(typefaceLoader, font)
+
+    fun setSubtitleTypeface(typefaceLoader: TypefaceLoader, font: FontOptions) = this.titleSubTitleBar.setSubtitleTypeface(typefaceLoader, font)
+
+    fun setSubtitleFontSize(size: Float) = this.titleSubTitleBar.setSubtitleFontSize(size)
+
+    fun setSubtitleColor(@ColorInt color: Int) = this.titleSubTitleBar.setSubtitleTextColor(color)
+
+    fun setTitleColor(@ColorInt color: Int) = this.titleSubTitleBar.setTitleTextColor(color)
+
+    fun getTitle(): String = this.titleSubTitleBar.getTitle()
+
+    fun clearCurrentTitle() {
+        this.titleSubTitleBar.clear()
+        clearComponent()
+    }
+
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        val titleComponent = getTitleComponent()
+        val isCenter = titleComponentAlignment == Alignment.Center
+        val containerWidth = r - l
+        val containerHeight = b - t
+        val isRTL = isRTL()
```

Comment:
since I'm calling the same function in multiple places in `onLayout` its better to store the same value and make less function calls, still readable.

Reply:
All `isRtl()` does is compare a value. You can safely assume that calling it twice has absolutely zero performance implications.

Applications running RNN are installed on state of the art, 64bit devices with plenty of memory and multiple cores - they can handle comparing to integers.

Reply:
I see that this is clearer and consistent with the nearby values I use. 
Since we both know device's capabilities (thanks for your reminders), we can agree that my way and your way are readable, straightforward, and have the same result for each run.
So it is just a matter of preferred style.

Reply:
I'm glad we agree that invoking `isRtl()` more than once has no performance implications. This leads me to something I noticed in your PRs - at times it seems like a lot of preparation work is done before calling some functions.
Just here, 8 lines are required to prepare various parameters for `resolveHorizontalTitleBoundsLimit()`.

Thankfully, Android Studio supports named arguments so we no longer need to declare parameters in advance for readability. For example, previously passing null and false was very confusing. We had to declare variables for the code to be readable. With named variables, this isn't needed anymore. The IDE provides named arguments when needed as visual aids.
<img width="737" alt="Screen Shot 2021-05-12 at 10 47 06" src="https://user-images.githubusercontent.com/12352397/117938280-87410a00-b30f-11eb-9c76-0150a9994882.png">

I strongly feel you should inline these variables. Once you do, you'll notice you can get rid of some of the parameters and calculate them in the functions themselves.

* `isCenter` is inferred from `titleComponentAlignment` which is a public member of `TitleBarAndButtonsContainer`.
* isRtl is now a public extension function of View which in our case is `TitleBarAndButtonsContainer`
* `leftBarWidth` is inferred from `leftButtonBar` which is public in `TitleBarAndButtonsContainer`
* `rightBarWidth` is inferred from `rightButtonBar` which is public in `TitleBarAndButtonsContainer`
* `titleWidth` is inferred from `titleComponent` which is public in  `TitleBarAndButtonsContainer`
* `titleComponent` is the result of the internal function `getTitleComponent()` - it can be made public.

You can pass **`this`**, `r`, `l` as parameters to `resolveHorizontalTitleBoundsLimit`. In `resolveHorizontalTitleBoundsLimit` use the `with(titleAndButtonsContainer)` so you can easily access whatever public members/getters you need.

Instead of 6 parameters, we're down to 3 and we can safely avoid the 8 line preparation work. An additional benefit is that calculations (like `containerWidth` and `containerHeight `) are done closer to where they are used.

Reply:
You have a good point here dude :] 

IDEs today are really helpful as you described, but GitHub does not have such a feature, now,  what these preps do is making measurement logic independent from any `View` (android specific code) and use just numbers, which means, writing simple and short tests for the calculations without mocking `View` and mock return for each param you listed above. 

That was the only concern about having them like this, so changing it now would take a bit of time re-writing tests and organizing mocks.